src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,257 @@
+        /// <returns>The target object instance, if present.</returns>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal object? UnsafeGetTarget()
+        {
+            return nGetPrimary(_handle);
+        }
+
+        /// <summary>
+        /// Sets the target object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetTarget(object? target)
+        {
+            nSetPrimary(_handle, target);
+        }
+
+        /// <summary>
+        /// Sets the dependent object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Dependent"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetDependent(object? dependent)
+        {
+            nSetSecondary(_handle, dependent);
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <param name="dependent">The dependent instance, if available.</param>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <remarks>
+        /// This method mirrors <see cref="GetTargetAndDependent"/>, but without the allocation check.
+        /// The signature is also kept the same as the one for the internal call, to improve the codegen.
+        /// </remarks>
+        internal object? UnsafeGetTargetAndDependent(out object? dependent)
+        {
+            return nGetPrimaryAndSecondary(_handle, out dependent);
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        public void Dispose()
+        {
+            // Forces dependentHandle back to non-allocated state (if not already there)
+            // and frees the handle if needed.
+
+            if (_handle != IntPtr.Zero)
+            {
+                IntPtr handle = _handle;
+                _handle = IntPtr.Zero;
+                nFree(handle);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern IntPtr nInitialize(object? primary, object? secondary);
+
+#if DEBUG
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object? nGetPrimary(IntPtr dependentHandle);
+#else
+        private static unsafe object? nGetPrimary(IntPtr dependentHandle)
+        {
+            // This optimization is the same that is used in GCHandle in RELEASE mode.
+            // This is not used in DEBUG builds as the runtime performs additional checks.
+            // This conversion works because on CoreCLR, DependentHandle is really just a
+            // OBJECTHANDLE value, the same that GCHandle uses. They are defined as follows:
+            //
+            // struct OBJECTHANDLE__
```

Comment:
I appreciate the long comment, but I do not think that it is useful to have a long explanation how GC handles are implemented here. Details like this tend to get stale.

I think it would be sufficient to say that this is inlined copy of the logic from the `ObjectFromHandle` method in the unmanaged runtime.

Reply:
Sure! I've updated the comment in 641b4052d20e100bb6a74dac7f1f707a5742e808 as suggested 🙂